hail/python/hailtop/httpx.py

```diff
@@ -24,34 +24,241 @@ def client_session(*args, **kwargs) -> aiohttp.ClientSession:
+        return self.ws.get_extra_info(name, default)
+
+    def exception(self) -> Optional[BaseException]:
+        return self.ws.exception()
+
+    def ping(self, message: bytes = b'') -> None:
+        async_to_blocking(self.ws.ping(message))
+
+    def pong(self, message: bytes = b'') -> None:
+        async_to_blocking(self.ws.pong(message))
+
+    def send_str(self, data: str,
+                 compress: Optional[int] = None) -> None:
+        return async_to_blocking(self.ws.send_str(data, compress))
+
+    def send_bytes(self, data: bytes,
+                   compress: Optional[int] = None) -> None:
+        return async_to_blocking(self.ws.send_bytes(data, compress))
+
+    def send_json(self, data: Any,
+                  compress: Optional[int] = None,
+                  *, dumps: aiohttp.typedefs.JSONEncoder = aiohttp.typedefs.DEFAULT_JSON_ENCODER) -> None:
+        return async_to_blocking(self.ws.send_json(data, compress, dumps=dumps))
+
+    def close(self, *, code: int = 1000, message: bytes = b'') -> bool:
+        return async_to_blocking(self.ws.close(code=code, message=message))
+
+    def receive(self, timeout: Optional[float] = None) -> aiohttp.WSMessage:
+        return async_to_blocking(self.ws.receive(timeout))
+
+    def receive_str(self, *, timeout: Optional[float] = None) -> str:
+        return async_to_blocking(self.ws.receive_str(timeout=timeout))
+
+    def receive_bytes(self, *, timeout: Optional[float] = None) -> bytes:
+        return async_to_blocking(self.ws.receive_bytes(timeout=timeout))
+
+    def receive_json(self,
+                     *, loads: aiohttp.typedefs.JSONDecoder = aiohttp.typedefs.DEFAULT_JSON_DECODER,
+                     timeout: Optional[float] = None) -> Any:
+        return async_to_blocking(self.ws.receive_json(loads=loads, timeout=timeout))
+
+    def __iter__(self) -> 'BlockingClientWebSocketResponse':
+        return self
+
+    def __next__(self) -> aiohttp.WSMessage:
+        try:
+            return async_to_blocking(self.ws.__anext__())
+        except StopAsyncIteration as exc:
+            raise StopIteration() from exc
+
+
+T = TypeVar('T')  # pylint: disable=invalid-name
+U = TypeVar('U')  # pylint: disable=invalid-name
+
+
+class AsyncToBlockingContextManager(Generic[T, U]):
+    def __init__(self, context_manager, wrap: Callable[[T], U]):
+        self.context_manager = context_manager
+        self.wrap = wrap
+
+    def __enter__(self) -> U:
+        return self.wrap(async_to_blocking(self.context_manager.__aenter__()))
+
+    def __exit__(self,
+                 exc_type: Optional[Type[BaseException]],
+                 exc: Optional[BaseException],
+                 tb: Optional[TracebackType]) -> None:
+        async_to_blocking(self.context_manager.__aexit__(exc_type, exc, tb))
+
+
+class BlockingClientResponseContextManager(AsyncToBlockingContextManager):
+    def __init__(self, context_manager):
+        super().__init__(context_manager, BlockingClientResponse)
```

Comment:
BlockingClientResponse isn't Callable[[T], U]. How does this work?

Reply:
It is a Callable, the constructor takes exactly one argument. You can verify this with `mypy` and the the magic `reveal_type` command:
```
(base) # cat foo.py
from hailtop.httpx import BlockingClientResponse

reveal_type(BlockingClientResponse)
(base) # mypy foo.py
foo.py:3: note: Revealed type is 'def (client_response: aiohttp.client_reqrep.ClientResponse) -> hailtop.httpx.BlockingClientResponse'
```